core/services/workflows/v2/time_provider.go

```diff
@@ -1,15 +1,52 @@
 package v2
 
-import "time"
+import (
+	"time"
 
-// TODO: https://smartcontract-it.atlassian.net/browse/CAPPL-903 implement the methods
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/wasm/workflowLib"
+)
 
-type TimeProvider struct{}
+type TimeProvider struct {
+	workflowExecutionID string
+	timeSeqNum          int
+	donTimeStore        *workflowLib.DonTimeStore
+}
+
+func NewTimeProvider(workflowExecutionID string) TimeProvider {
+	return TimeProvider{
+		workflowExecutionID: workflowExecutionID,
+		timeSeqNum:          0,
+		donTimeStore:        workflowLib.GetDonTimeStore(),
+	}
+}
 
 func (tp *TimeProvider) GetNodeTime() time.Time {
-	return time.Now().UTC()
+	return fromUnixMilli(tp.donTimeStore.GetLastObservedDonTime())
 }
 
+// GetDONTime makes a request to the WorkflowLib plugin store for DON Time
 func (tp *TimeProvider) GetDONTime() time.Time {
```

Comment:
Don't we need a context passed here? How do we cancel this blocking call on Engine shutdown or workflow max execution time expiry?